doc/rtd/explanation/failure_states.rst

```diff
@@ -0,0 +1,138 @@
+.. _failure_states:
+
+Failure states: recoverable errors vs non-recoverable errors
+============================================================
+
+critical failure
+----------------
+If cloud-init is unable to complete, error messages will now
+additionally be visible in output of `cloud-init status --format json`
+within the 'error' key nested under the module-level keys: 'init-local',
+'init', 'modules-config', 'modules-final'.
+
+recoverable failure
+-------------------
+In the case that cloud-init is able to complete yet something goes awry,
+the service will now return with exit code 2, and error messages will be
+visible in the output of `cloud-init status --format` json under the top
+level 'recoverable_errors' key as well as within the 'error' key nested
+under the module-level keys: 'init-local', 'init', 'modules-config',
+'modules-final'.
+
+Implementation
+==============
+
+Cloud-init error codes
+----------------------
+ 0 - success
+ 1 - unrecoverable error
+ 2 - recoverable error
+
+If cloud-init exits with exit code 1, cloud-init experienced critical failure
+and was unable to recover. In this case, something is likely seriously
+wrong with the system, or cloud-init has experienced a serious bug.
+
+If cloud-init exits with exit code 2, cloud-init was able to complete
+gracefully, however something went wrong and the user should investigate.
+
+
+Reported state
+--------------
+Cloud-init can report its internal state via the `status --format json`
+subcommand undert the `'extended_status'` key.
+
+$ cloud-init status --format json | jq .extended_status
+"degraded done"
+
+See the list of all possible states:
+
+.. code-block: shell-session
```

Comment:
Having an inline list here but then appendices for the next two sections is a little odd, especially when this is the longest list. Personally, I prefer them all inline, but I'm fine either way as long as we are consistent.